src/MuTalk-Model/MutationTestingAnalysisBuilder.class.st

```diff
@@ -0,0 +1,136 @@
+Class {
+	#name : 'MutationTestingAnalysisBuilder',
+	#superclass : 'Object',
+	#instVars : [
+		'modelClasses',
+		'operators',
+		'testCases',
+		'logger',
+		'budget',
+		'testSelectionStrategy',
+		'mutantSelectionStrategy'
+	],
+	#category : 'MuTalk-Model',
+	#package : 'MuTalk-Model'
+}
+
+{ #category : 'accessing' }
+MutationTestingAnalysisBuilder >> budget: anObject [
+
+	budget := anObject
+]
+
+{ #category : 'builder' }
+MutationTestingAnalysisBuilder >> buildAnalysis [
+
+	^ MutationTestingAnalysis new
+		  initializeFor: testCases
+		  mutating: modelClasses
+		  using: operators
+		  with: testSelectionStrategy
+		  with: mutantSelectionStrategy
+		  loggingIn: logger
+		  budget: budget
+]
+
+{ #category : 'accessing' }
+MutationTestingAnalysisBuilder >> classesToMutate: anObject [
+
+	modelClasses := anObject
+]
+
+{ #category : 'accessing - defaults' }
+MutationTestingAnalysisBuilder >> defaultBudget [
+
+	^ MTFreeBudget new
+]
+
+{ #category : 'defaults' }
+MutationTestingAnalysisBuilder >> defaultLogger [
+
+	^ NullMutationTestingAnalysisLogger new
+]
+
+{ #category : 'accessing - defaults' }
+MutationTestingAnalysisBuilder >> defaultMutantSelectionStrategy [
+
+	^ AllMutantSelectionStrategy new
+]
+
+{ #category : 'accessing - defaults' }
+MutationTestingAnalysisBuilder >> defaultOperators [
+
+	^ MutantOperator contents
+]
+
+{ #category : 'accessing - defaults' }
+MutationTestingAnalysisBuilder >> defaultTestSelectionStrategy [
+
+	^ AllTestsMethodsRunningTestSelectionStrategy new
+]
+
+{ #category : 'initialization' }
+MutationTestingAnalysisBuilder >> initialize [
+
+	super initialize.
+	budget := self defaultBudget.
+	logger := self defaultLogger.
+	mutantSelectionStrategy := self defaultMutantSelectionStrategy.
+	operators := self defaultOperators.
+	testSelectionStrategy := self defaultTestSelectionStrategy
+]
+
+{ #category : 'accessing' }
+MutationTestingAnalysisBuilder >> logger: anObject [
+
+	logger := anObject
+]
+
+{ #category : 'accessing' }
+MutationTestingAnalysisBuilder >> mutantSelectionStrategy: anObject [
+
+	mutantSelectionStrategy := anObject
+]
+
+{ #category : 'accessing' }
+MutationTestingAnalysisBuilder >> operators: anObject [
+
+	operators := anObject
+]
+
+{ #category : 'accessing' }
+MutationTestingAnalysisBuilder >> testCases: anObject [
+
+	testCases := anObject
+]
+
+{ #category : 'accessing' }
+MutationTestingAnalysisBuilder >> testCasesFrom: aClassCollection [
+
+	^ aClassCollection
+		  inject: OrderedCollection new
+		  into: [ :testCase :testClass |
+			  ((testClass isKindOf: TestCase class) and: [
```

Comment:
This condition is strange. This means that if I pass as argument a class that is not a test class, the builder will *silently fail*. I mean, it will ignore my class, but will not tell me why, it will not tell me I was using it wrong!

This could make people spend hours in silly debugging saying "Why is it not working?!!".

I would prefer that
 - either we validate that all test classes are test classes and throw an exception if not (and maybe we need a test for this)
 - or we trust the client and do not validate the types (and this it will throw a DNU when not well used)

@DurieuxPol check
 - http://rmod-pharo-mooc.lille.inria.fr/AdvancedDesignMooc/Slides/M5-7-Principles-01-AvoidIsNil.pdf
 - http://rmod-pharo-mooc.lille.inria.fr/AdvancedDesignMooc/Videos/M05_S7_V2.mp4